linera-witty/src/memory_layout/split_flat_layouts.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Splitting of the flat layout of a `variant` type into the flat layout of one of its variants.
+//!
+//! When flattening `variant` types, a single flat layout must be obtained for the type by joining
+//! the flat layout of each variant. This means finding a flat type for each layout element to
+//! represent the flat type of any of the variants. See [`crate::primitive_types::JoinFlatTypes`]
+//! for more information on how flat types are joined.
+
+use crate::primitive_types::FlatType;
+use frunk::{HCons, HNil};
+
+/// Converts the current joined flat layout into the `Target` flat layout, which may be shorter or
+/// have elements that are narrower than the current elements.
+pub trait SplitFlatLayouts<Target> {
+    /// Converts the current joined flat layout into the `Target` flat layout.
+    fn split(self) -> Target;
```

Comment:
nit: similarly, this name doesn't convey the notion of conversion (in the opposite direction).

Reply:
Done, it's now `JoinFlatLayouts::from_joined(_: Target) -> Self`